### DISCUSSION TITLE
Rationale behind `PartialMessage` and, in particular, allowing it as a return for method implementations

### DISCUSSION THREAD

**Discussion by fubhy** [connectrpc/connect-es#348] [Q&A]
I am currently wondering whether allowing `PartialMessage` to be returned from service methods is a good idea or not. By adding ambiguouity there, it makes it harder to implement generic decorators / wrappers for service methods that could otherwise easily deal with the returned messages.

For instance, a generic logging or caching wrapper that wraps `MethodImpl` now needs to check the returned value from the underlying implementation and possibly cast it into a proper message.

I think it would be a better developer experience overal if service method implementations were strictly required to return proper messages.

Or is there a reason why this is allowed currently?

**Comment by fubhy** [comment]
And, if that's intentional and desirable, I'd suggest we add a public `normalize` function to `connect-node` to make that more convenient?

**Comment by fubhy** [comment]
E.g. I currently need to do sth. like this:

```typescript
import type { AnyMessage, Message, MessageType, PartialMessage } from '@bufbuild/protobuf';
import md5 from 'md5';

import type { EnhancedContext } from './context';
import type { UnaryImpl } from './service';

export function cacheServiceMethod<TInput extends Message<TInput>, TOutput extends Message<TOutput>>(
  implementation: UnaryImpl<TInput, TOutput, EnhancedContext>,
  ttl: number | string = '1 minute',
): UnaryImpl<TInput, TOutput, EnhancedContext> {
  return async (request, ctx) => {
    const cachePrefix = `grpc.${ctx.context.service.typeName}.${ctx.context.method.name}`;
    const cacheSuffix = md5(request.toBinary({ writeUnknownFields: false }));
    const cacheKey = `${cachePrefix}:${cacheSuffix}`;

    const cacheValue = await ctx.cache.getBuffer(cacheKey);
    const messageType = ctx.context.method.O as MessageType<TOutput>;

    if (cacheValue !== null) {
      return messageType.fromBinary(cacheValue, { readUnknownFields: false });
    }

    const methodResult = normalize(await implementation(request, ctx), messageType);
    const methodResultBinary = methodResult.toBinary({ writeUnknownFields: false });

    await ctx.cache.setBuffer(cacheKey, Buffer.from(methodResultBinary), ttl);

    return methodResult;
  };
}

function normalize<TMessage extends Message<TMessage> = AnyMessage>(
  output: PartialMessage<TMessage> | TMessage,
  type: MessageType<TMessage>,
): TMessage {
  return output instanceof type ? output : new type(output as PartialMessage<TMessage>);
}
```

**Reply by timostamm** [reply]
The partial return is convenient for implementations. For example, this one for [eliza](https://github.com/bufbuild/connect-demo/blob/main/proto/buf/connect/demo/eliza/v1/eliza.proto):

```ts
  say(req) {
    return {
       sentence: `you said: ${req.sentence}`
    }
  }
```

It's equivalent to clients, where you can do:

```ts
client.say({ sentence: "hello" });
```

I think this is a desirable feature. 

Have you used interceptors (for clients / Transport) yet? 

Thanks for the example code, use cases like this are awesome!

**Reply by fubhy** [reply]
Ok. I think I can buy into that. Especially for the client... But maybe this could be optional?

I'm still not sure whether there's the same benefit on the server or not. Given that there you are the developer choosing to use gRPC and there's not that much overhead to typing out `new MyMessage(...)` either. Also, there's arguably stronger typechecking too if you explicitly enforce a message type instance instead of a property bag that typescript only loosely checks for whether it satisfied the properties of the expected object.

E.g. because of `PartialMessage`, you might actually miss bugs in your code. Yesterday I was doing a lot of copy & pasting trying out a bunch of different things and ended up with a wrong message type returned from one of my methods. But because of the loose matching of `PartialMessage`, typescript didn't complain.

How about we make this an optional type parameter that, if the service / client is configured as "loose" (through an option during instantiation), we allow `Message<I / O> | PartialMessage<I / O>` instead of a full `Message<I / O>`?

**Reply by fubhy** [reply]
Another problem imho is that this is also part of `protoc-gen-es` and allows nested messages to also be set as `PartialMessage`. That means that there's actually no types enforced there and you could accidentlly set a completely different message.

E.g. for a schema like ...


```proto
message Foo {
  Bar bar = 1;
}

message Bar {
  string foo = 1;
  string bar = 2;
}

message Baz {
  string baz = 1
}
```

Typescript wouldn't complain if you did this:

```typescript
const foo = new Foo({ bar: new Baz({ baz: "hello world!" ) });
```

So I'd personally prefer it to be optional there too actually (configurable in protoc-gen-es?). I'd personally always side with "strict types by default" and then, e.g. allow the client to be configured with e.g. `strict: false` which then results in the client becoming a `LooselyTypedClient<T>` that recursively drops in the `PartialMessage` wrappers :-).

**Reply by fubhy** [reply]
And sorry for all this early stage noise in your repository. I'm just so excited about grpc finally seeing the love it deserves also in the browser :-)

**Reply by timostamm** [reply]
An early version of connect-web offered two ways to create a promise-client. One with partial input, one with non-partial. You can still see it [here](https://github.com/bufbuild/connect-web/blob/ed409605a2ed66f738f8247fdef6c435d0bdb196/packages/connect-web/src/promise-client.ts). We removed the feature to keep the code as simple as possible, but obviously users can still roll their own client to their preference. 

For the server-side, we would ideally offer similar functionality, but proper middleware has higher priority. Similar to interceptors, users could do: 
```ts
createHandler(someService, someMethod, methodImpl, {
  middleware: [ cacheServiceMethod("1 minute") ]
})
```

This makes them nicely re-usable. But interceptors and middleware needs to be able to work with any method - unary and all stream types. Unfortunately, this is a bit tricky in TypeScript, because you can't differentiate between a union of function signatures, so we need intermediate types with a discriminator property. The approach you're taking above is great, but it only works for unary, and most of the time, you want the authentication middleware to work for every RPC.

We didn't originally plan to add this feature in the first version of connect-node, but I think we need to, just to remove the large amounts of duplicate code in the protocol handlers.

But to get back to the partial input - did you see [function unimplementService()](https://github.com/bufbuild/connect-web/blob/1fbecc1252ad9532aba0aa2df93ec29e8495e369/packages/connect-node/src/implementation.ts#L40)? It wraps a service implementation that may be missing some methods, and provides a fallback for them. I think there's a bit of an overlap between this function and the partial outputs. I'm wondering if something else will come up we haven't thought about yet. Does seem a bit early to take action now, considering that this hasn't seen much usage yet.

**Reply by fubhy** [reply]
> This makes them nicely re-usable. But interceptors and middleware needs to be able to work with any method - unary and all stream types. Unfortunately, this is a bit tricky in TypeScript, because you can't differentiate between a union of function signatures, so we need intermediate types with a discriminator property. The approach you're taking above is great, but it only works for unary, and most of the time, you want the authentication middleware to work for every RPC.

The way I "resolved" that for us is by simply having different wrappers for different method kinds. Not great in terms of developer experience & code reuse, but works.

> But to get back to the partial input - did you see [function unimplementService()](https://github.com/bufbuild/connect-web/blob/1fbecc1252ad9532aba0aa2df93ec29e8495e369/packages/connect-node/src/implementation.ts#L40)?

Yeah, I'm using that already and took a similar route to add a global / generic wrapper around all service methods to inject even more stuff into the `context` (service dependencies, database, etc.). For static stuff, that's not *really* necessary because you could obviously just inject that through a `creator` function that returns a `MethodImpl` but it's how I am trying to do auth (every method needs auth in my setup).... Since auth is request-context dependent (api token in header).

I think there's a need for service-level middlewares / interceptors and method-level middlewares / interceptors. Would you generally agree with that or just put everything into method-level (and accept the small amount of duplication you'll end up with there)?

--------------------------

Back to partial messages:

> An early version of connect-web offered two ways to create a promise-client. One with partial input, one with non-partial. You can still see it [here](https://github.com/bufbuild/connect-web/blob/ed409605a2ed66f738f8247fdef6c435d0bdb196/packages/connect-web/src/promise-client.ts). We removed the feature to keep the code as simple as possible, but obviously users can still roll their own client to their preference.

I think my objection regarding PartialMessage is simply rooted in the "loose typing by default". I can see how this is useful especially for client convenience. But I'd have personally opted for a "strict by default & loose by opt-in" approach there. E.g. by spinning up a `LooseClient` instance that wraps all request input for all methods in the `PartialMessage` type and thereby recursively applies to also to nested messages. I think I'd personally also prefer not having `PartialMessage` in my generated types from `protoc-gen-es` by default. Especially considering how that can be easily retro-fitted manually by simply wrapping that type around a message after-the-fact. It's certainly harder to do the opposite and eliminate the PartialMessage type from a message through some typescript type inference magic (which is possible, but not as convenient). In such a case, you'd also weirdly end up with a construct like `StrictMessage<PartialMessage<MyMessage>>`. 

Again, I can see the utility of PartialMessage for certain client use-cases, especially for simple demos or for external users. But in-house, I'd always opt for the strict types and I'd imagine & hope that most teams would (Internally). The added "peace of mind" especially during larger refactors, etc. is certainly worth the few extra keystrokes. 

> Does seem a bit early to take action now, considering that this hasn't seen much usage yet.

Agreed. With everything I'm posting here I'm merely documenting my journey and noting down things that crossed my mind along the way. I hope that's helpful but feel free to ignore it ;-)

**Reply by timostamm** [reply]
> I think there's a need for service-level middlewares / interceptors and method-level middlewares / interceptors. Would you generally agree with that

Apologies for the late reply. With clients, I don't think we'll add _method-level_ interceptors for now. It's pretty straight-forward to exempt an interceptor from a specific method, for example:

```ts
function authHeader(... except: MethodInfo[]): Interceptor {
  return (next) => async (req) => {
    if (except.includes(req.method)) {
      return await next(req);
    }
    req.header.set("X-Token", "foo");
    return await next(req);
  };
}

createConnectTransport({
  baseUrl: "https://example.com",
  interceptors: [
    authHeader(TestService.methods.fullDuplexCall)
  ]
})
```

For servers, interceptors are very useful, but we wanted to see how far we can go without them. That said, it currently looks like we're going to add them. Similar to clients, they will apply to every RPC. But we'll certainly accept them in `function createHandler()`, so if you pass an interceptor there, it only applies to the one RPC. I'm not sure yet what we're going to do with the context.

**Reply by timostamm** [reply]
Very interesting feedback on `PartialMessage`, thank you!

**Reply by fubhy** [reply]
For servers, I think per-method middlewares could be rather useful.

For instance, what about method-specific cache ttl, rate limiting constraints, auth requirements, etc.?

I agree that this seems to be less useful for the client.
